langchain4j-google-ai-gemini/src/main/java/dev/langchain4j/model/googleai/GoogleAiGeminiStreamingChatModel.java

```
@@ -0,0 +1,228 @@
+                        .temperature(this.temperature)
+                        .topK(this.topK)
+                        .topP(this.topP)
+                        .build())
+                .safetySettings(this.safetySettings)
+                .tools(FunctionMapper.fromToolSepcsToGTool(toolSpecifications, this.allowCodeExecution))
+                .toolConfig(new GeminiToolConfig(this.toolConfig))
+                .build();
+    }
+
+    private ChatModelRequest createChatModelRequest(List<ChatMessage> messages, List<ToolSpecification> toolSpecifications) {
+        return ChatModelRequest.builder()
+                .model(modelName)
+                .temperature(temperature)
+                .topP(topP)
+                .maxTokens(maxOutputTokens)
+                .messages(messages)
+                .toolSpecifications(toolSpecifications)
+                .build();
+    }
+
+    private void processGenerateContentRequest(GeminiGenerateContentRequest request, StreamingResponseHandler<AiMessage> handler,
+                                               ChatModelRequest chatModelRequest, ConcurrentHashMap<Object, Object> listenerAttributes) {
+        GeminiStreamingResponseBuilder responseBuilder = new GeminiStreamingResponseBuilder(this.includeCodeExecutionOutput);
+
+        try {
+            Stream<GeminiGenerateContentResponse> contentStream = withRetry(
+                    () -> this.geminiService.generateContentStream(this.modelName, this.apiKey, request),
+                    maxRetries);
+
+            contentStream.forEach(partialResponse -> {
+                Optional<String> text = responseBuilder.append(partialResponse);
+                text.ifPresent(handler::onNext);
+            });
+
+            Response<AiMessage> fullResponse = responseBuilder.build();
+            handler.onComplete(fullResponse);
+
+            notifyListenersOnResponse(fullResponse, chatModelRequest, listenerAttributes);
+        } catch (RuntimeException exception) {
```

What if this is not a runtime exception ?

---

Sure, I can simply catch all exceptions; I based this on the GoogleAiGeminiCatModel class, which also only catches RuntimeExceptions. What other exceptions could even be thrown here? I thought I’m actually already catching all the others beforehand, aren’t I?

---

@kpavlov what else can this be?

---

Maybe Langchain4jException extends RuntimeException. Not important

---

We plan to intoroduce a set of common exceptions for common situations (e.g. bad request, rate limit, service unavailable, etc), so let's not add anything now that has to be changed again soon